builds/jazz-build-module/azure-function-deploy-module.groovy

```diff
@@ -0,0 +1,308 @@
+        assetType = "eventhubs_namespace"
+        break
+      case "Microsoft.DocumentDB/databaseAccounts":
+        assetType = "cosmosdb_account"
+        break
+    }
+
+    events.sendCompletedEvent('CREATE_ASSET', null, utilModule.generateAssetMap(serviceInfo.serviceCatalog['platform'], id, assetType, serviceInfo.serviceCatalog), serviceInfo.envId)
+
+  }
+
+  if (assetList) {
+    for (item in assetList) {
+      def id = item.azureResourceId
+      def assetType = item.type
+      events.sendCompletedEvent('CREATE_ASSET', null, utilModule.generateAssetMap(serviceInfo.serviceCatalog['platform'], id, assetType, serviceInfo.serviceCatalog), serviceInfo.envId)
+
+    }
+  }
+}
+
+def invokeAzureCreation(serviceInfo){
+
+  sh "rm -rf _azureconfig"
+
+  sh "zip -qr content.zip ."
+  def zip = sh(script: 'readlink -f ./content.zip', returnStdout: true).trim()
+
+  withCredentials([
+    [$class: 'UsernamePasswordMultiBinding', credentialsId: 'AZ_PASSWORD', passwordVariable: 'AZURE_CLIENT_SECRET', usernameVariable: 'UNAME'],
+    [$class: 'UsernamePasswordMultiBinding', credentialsId: 'AZ_CLIENTID', passwordVariable: 'AZURE_CLIENT_ID', usernameVariable: 'UNAME'],
+    [$class: 'UsernamePasswordMultiBinding', credentialsId: 'AZ_TENANTID', passwordVariable: 'AZURE_TENANT_ID', usernameVariable: 'UNAME'],
+    [$class: 'UsernamePasswordMultiBinding', credentialsId: 'AZ_SUBSCRIPTIONID', passwordVariable: 'AZURE_SUBSCRIPTION_ID', usernameVariable: 'UNAME']
+  ]) {
+
+    def type = azureUtil.getExtensionName(serviceInfo)
+    def runtimeType = azureUtil.getRuntimeType(serviceInfo)
+    def tags = azureUtil.getTags(serviceInfo)
+    def data = azureUtil.getAzureRequestPayload(serviceInfo)
+    data.tags = tags
+    data.runtime = runtimeType
+    data.resourceName = serviceInfo.resourceName
+
+
+    def repo_name = "jazz_azure-create-service"
+    sh "rm -rf $repo_name"
+    sh "mkdir $repo_name"
+
+    def repocloneUrl = scmModule.getCoreRepoCloneUrl(repo_name)
+    def masterKey
+
+    dir(repo_name)
+      {
+        def assetList =[]
+        checkout([$class: 'GitSCM', branches: [[name: '*/master']], doGenerateSubmoduleConfigurations: false, extensions: [], submoduleCfg: [], userRemoteConfigs: [[credentialsId: configLoader.REPOSITORY.CREDENTIAL_ID, url: repocloneUrl]]])
+        sh "npm install -s"
+        try {
+
+          createStorageAccount(data, serviceInfo)
+
+          if (type) {
+            assetList = createEventResource(data, type, serviceInfo)
+          }
+
+          createFunctionApp(data)
+
+          def output = azureUtil.invokeAzureService(data, "getMasterKey")
+          masterKey = output.data.result.key
+          deployFunction(data, zip, type)
+          sendAssetCompletedEvent(serviceInfo, assetList)
+          return masterKey
+        } catch (ex) {
+          echo "error occur $ex, rollback starting..."
+          deleteResourceByTag(serviceInfo)
+          error "Failed creating azure function $ex"
```

Comment:
Every error (jenkins error) should be caught and FAILURE events must be sent before the jenkins workflow errors out. Unless this happens, we can't mark the deployment as failed in jazz metadata.